src/index/TextIndexReadWrite.h

```diff
@@ -0,0 +1,156 @@
+  LOG(DEBUG) << "Decoding Simple8b code...\n";
+  ad_utility::Simple8bCode::decode(encoded, nofElements,
+                                   frequencyEncodedResult.data());
+  LOG(DEBUG) << "Reverting frequency encoded items to actual IDs...\n";
+  frequencyEncodedResult.resize(nofElements);
+  vector<To> result;
+  result.reserve(frequencyEncodedResult.size());
+  for (size_t i = 0; i < frequencyEncodedResult.size(); ++i) {
+    result.push_back(transformer(codebook[frequencyEncodedResult[i]]));
+  }
+  delete[] encoded;
+  delete[] codebook;
+  LOG(DEBUG) << "Done reading frequency-encoded list. Size: " << result.size()
+             << "\n";
+  return result;
+}
+
+template <typename To, typename From>
+vector<To> readGapComprList(
+    size_t nofElements, off_t from, size_t nofBytes,
+    const ad_utility::File& textIndexFile,
+    const std::function<To(From)>& transformer = [](From x) {
+      return static_cast<To>(x);
+    }) {
+  LOG(DEBUG) << "Reading gap-encoded list from disk...\n";
+  LOG(TRACE) << "NofElements: " << nofElements << ", from: " << from
+             << ", nofBytes: " << nofBytes << '\n';
+  vector<From> gapEncodedVector;
+  gapEncodedVector.resize(nofElements + 250);
+  uint64_t* encoded = new uint64_t[nofBytes / 8];
+  textIndexFile.read(encoded, nofBytes, from);
+  LOG(DEBUG) << "Decoding Simple8b code...\n";
+  ad_utility::Simple8bCode::decode(encoded, nofElements,
+                                   gapEncodedVector.data());
+  LOG(DEBUG) << "Reverting gaps to actual IDs...\n";
+  gapEncodedVector.resize(nofElements);
+
+  // Undo gapEncoding
+  vector<To> result;
+  result.reserve(nofElements);
+  From previous = 0;
+  for (size_t i = 0; i < gapEncodedVector.size(); ++i) {
+    previous += gapEncodedVector[i];
+    result.push_back(transformer(previous));
+  }
+  delete[] encoded;
+  LOG(DEBUG) << "Done reading gap-encoded list. Size: " << result.size()
+             << "\n";
+  return result;
+}
+
+}  // namespace TextIndexReadWrite
+
+template <typename T>
+class FrequencyEncode {
+ public:
+  using TypedMap = ad_utility::HashMap<T, size_t>;
+  using TypedVector = std::vector<T>;
+  using CodeMap = TypedMap;
+  using CodeBook = TypedVector;
+
+  explicit FrequencyEncode(const TypedVector& vectorToEncode);
+
+  void writeToFile(ad_utility::File& out, size_t nofElements,
+                   off_t& currentOffset);
+
+  const TypedVector getEncodedVector() { return encodedVector_; }
+  const CodeMap& getCodeMap() { return codeMap_; }
+  const CodeBook& getCodeBook() { return codeBook_; }
+
+ private:
+  std::vector<size_t> encodedVector_;
+  CodeMap codeMap_;
+  CodeBook codeBook_;
+};
+
+template <typename T>
+requires std::is_arithmetic_v<T> class GapEncode {
+ public:
+  using TypedVector = std::vector<T>;
+
+  explicit GapEncode(const TypedVector& vectorToEncode);
+
+  void writeToFile(ad_utility::File& out, size_t nofElements,
+                   off_t& currentOffset);
```

Comment:
1. I think you can get away with a lazy view as the input to the constructor.
2. Would the code in `IndexImpl.cpp` become simpler if you make a static function that does the encoding + writing in one step (same for the other encoders).

Reply:
(Also maybe part for a separate PR, your work here is valuable, by moving it to a separate file it now has a size where we can see the possible improvements much simpler.